File: src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/ExpandingRangeIntGenerator.java
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package org.jetbrains.kotlinx.lincheck.paramgen;

import java.util.Random;

/**
 * Represents a random number generator from specified extensible range.
 * Uses two ranges as parameters: initial and maximal range.
 * After each generation takes an arbitrary number from the current range and with 50% probability expands its
 * right or left bound. This process continues until current range bounds are equals to maximal range bounds.
 */
public class ExpandingRangeIntGenerator {
    private final Random random;

    private static final int DEFAULT_START_VALUE = 0;

    /**
     * Left (inclusive) bound of current range
     */
    private int begin;
    /**
     * Right (inclusive) bound of current range
     */
    private int end;

    /**
     * Minimal left (inclusive) bound of current range
     */
    private final int minBegin;
    /**
     * Maximal right (inclusive) bound of current range
     */
    private final int maxEnd;

    private NextExpansionDirection nextExpansionDirection = NextExpansionDirection.UP;

    /**
     * Creates extensible range-based random generator.
     *
     * @param random   random for this generator
     * @param begin    initial range left bound (inclusive)
     * @param end      initial range right bound (inclusive)
     * @param minBegin minimal range left bound (inclusive)
     * @param maxEnd   maximal range right bound (inclusive)
     */
    public ExpandingRangeIntGenerator(Random random, int begin, int end, int minBegin, int maxEnd) {
        this.random = random;
        this.begin = begin;
        this.end = end;
        this.minBegin = minBegin;
        this.maxEnd = maxEnd;
    }


    /**
     * Generates next random number.
     * With 50% percent probability expands current range if can and returns an expanded bound as a value.
     * Otherwise, returns a random value from the current range
     *
     * @return random value from current range or moved bound
     */
    public int nextInt() {
        checkRangeExpansionAbility();

        if (nextExpansionDirection == NextExpansionDirection.DISABLED || !random.nextBoolean()) {
            return generateFromRandomRange(begin, end);
        }

        if (nextExpansionDirection == NextExpansionDirection.DOWN) {
            nextExpansionDirection = NextExpansionDirection.UP;
            return --begin;
        } else {
            nextExpansionDirection = NextExpansionDirection.DOWN;
            return ++end;
        }
    }

    private void checkRangeExpansionAbility() {
        if (begin == minBegin && end == maxEnd) {
            nextExpansionDirection = NextExpansionDirection.DISABLED;
            return;
        }
        if (begin == minBegin) {
            nextExpansionDirection = NextExpansionDirection.UP;
            return;
        }
        if (end == maxEnd) {
            nextExpansionDirection = NextExpansionDirection.DOWN;
        }
    }

    private enum NextExpansionDirection {
        UP,
        DOWN,
        DISABLED
    }

    private int generateFromRandomRange(int rangeLowerBound, int rangeUpperBound) {
        return rangeLowerBound + random.nextInt(rangeUpperBound - rangeLowerBound + 1);
    }

    /**
     * Creates extensible range-based random generator.
     *
     * @param random        random for this generator
     * @param configuration range configuration in format begin:end, may be empty
     * @param minBegin      minimal range left bound (inclusive)
     * @param maxEnd        maximal range right bound (inclusive)
     * @param type          name of current type for generation for exception message if such appears
     * @return created range-based random generator§
     */
    public static ExpandingRangeIntGenerator create(Random random, String configuration, int minBegin, int maxEnd, String type) {
        if (configuration.isEmpty()) {
            return new ExpandingRangeIntGenerator(random, DEFAULT_START_VALUE, DEFAULT_START_VALUE, minBegin, maxEnd);
        }

        String[] args = configuration.replaceAll("\\s", "").split(":");
        if (args.length != 2) {
            throw new IllegalArgumentException("Configuration should have " +
                    "two arguments (begin and end) separated by colon");
        }

        // begin:end
        int begin = Integer.parseInt(args[0]);
        int end = Integer.parseInt(args[1]);
        if (begin < minBegin || end - 1 > maxEnd) {
            throw new IllegalArgumentException("Illegal range for "
                    + type + " type: [" + begin + "; " + end + ")");
        }

        int startValue = begin + (end - begin) / 2;
        if (end < begin) {
            throw new IllegalArgumentException("end must be >= than begin");
        }
        if (maxEnd < minBegin) {
            throw new IllegalArgumentException("maxEnd must be >= than minBegin");
        }
        if (end > maxEnd) {
            throw new IllegalArgumentException("end must be <= than maxEnd");
        }

        return new ExpandingRangeIntGenerator(random, startValue, startValue, begin, end);
    }


}

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/ExpandingRangeIntGenerator.kt
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>
 */

package org.jetbrains.kotlinx.lincheck.paramgen

import org.jetbrains.kotlinx.lincheck.paramgen.ExpandingRangeIntGenerator.NextExpansionDirection.*
import java.util.*


/**
 * Extensible range-based random generator.
 *
 * @param random   random for this generator
 * @param startInclusive    initial range left bound (inclusive)
 * @param endInclusive      initial range right bound (inclusive)
 * @param minStartInclusive minimal range left bound (inclusive)
 * @param maxEndInclusive   maximal range right bound (inclusive)
 */
class ExpandingRangeIntGenerator(
    private val random: Random,
    private val minValueInclusive: Int,
    private val maxValueInclusive: Int
) {

    /**
     * Current left (inclusive) bound of current range
     */
    private var currentStartInclusive: Int

    /**
     * Current right (inclusive) bound of current range
     */
    private var currentEndInclusive: Int

    private var nextExpansionDirection = UP

    init {
        reset()
        currentStartInclusive = 0
        currentEndInclusive = 0
    }

    /**
     * Reset bounds of this expanding range
     */
    fun reset() {
        val firstValue = ((minValueInclusive.toLong() + maxValueInclusive.toLong()) / 2).toInt()
        currentStartInclusive = firstValue
        currentEndInclusive = firstValue
    }

    /**
     * Generates next random number.
     * With 50% percent probability expands current range if can and returns an expanded bound as a value.
     * It alternates expansion direction from left to right.
     * Otherwise, returns a random value from the current range.
     *
     * @return random value from current range or moved bound
     */
    fun nextInt(): Int {
        if (nextExpansionDirection == DISABLED || random.nextDouble() > 0.5) {
            return generateFromRandomRange(currentStartInclusive, currentEndInclusive)
        }
        val value = if (nextExpansionDirection == DOWN) {
            --currentStartInclusive
        } else {
            ++currentEndInclusive
        }
        updateExpansionDirection()
        return value
    }

    private fun updateExpansionDirection() {
        if (currentStartInclusive == minStartInclusive && currentEndInclusive == maxEndInclusive) {
            nextExpansionDirection = DISABLED
            return
        }
        if (currentStartInclusive == minStartInclusive) {
            nextExpansionDirection = UP
            return
        }
        if (currentEndInclusive == maxEndInclusive) {
            nextExpansionDirection = DOWN
        }
    }

    private enum class NextExpansionDirection {
        UP,
        DOWN,
        DISABLED
    }

    private fun generateFromRandomRange(rangeLowerBoundInclusive: Int, rangeUpperBoundInclusive: Int): Int {
        return rangeLowerBoundInclusive + random.nextInt(rangeUpperBoundInclusive - rangeLowerBoundInclusive + 1)
    }

}

/**
 * Factory method to create extensible range with string configuration
 *
 * @param random            random for this generator
 * @param configuration     string configuration of format startInclusive:endInclusive
 * @param minStartInclusive minimal range left bound (inclusive)
 * @param maxEndInclusive   maximal range right bound (inclusive)
 * @param type              type of generator which is using this method to throw exception with this type name im message
 */
internal fun ExpandingRangeIntGenerator(
    random: Random, configuration: String, minStartInclusive: Int, maxEndInclusive: Int, type: String
): ExpandingRangeIntGenerator {
    if (configuration.isEmpty()) return ExpandingRangeIntGenerator(
        random = random,
        minStartInclusive = minStartInclusive,
        maxEndInclusive = maxEndInclusive
    )

    // start:end
    val args = configuration.replace("\\s", "").split(":")

    require(args.size == 2) { "Configuration should have two arguments (start and end) separated by colon" }

    val startInclusive = args[0].toIntOrNull() ?: error("Bad $type configuration. StartInclusive value must be a valid integer.")
    val endInclusive = args[1].toIntOrNull() ?: error("Bad $type configuration. EndInclusive value must be a valid integer.")
    require(startInclusive >= minStartInclusive || endInclusive - 1 <= maxEndInclusive) { "Illegal range for $type type: [$startInclusive; $endInclusive)" }

    require(maxEndInclusive >= minStartInclusive) { "maxEnd must be >= than minStart" }
    require(endInclusive <= maxEndInclusive) { "end must be <= than maxEnd" }

    return ExpandingRangeIntGenerator(
        random = random,
        minStartInclusive = startInclusive,
        maxEndInclusive = endInclusive
    )
}
